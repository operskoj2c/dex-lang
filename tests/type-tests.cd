:t lam x: x

:t lam x: sum for i: x.i

:t let flip f x y = f y x in flip

:t lam x: for i j: x.j.i

:t lam f x: f x

:t let id x = x
   in id (id) (id 1)

:t let x = unpack iota 10
   in sum x

:t let x = unpack iota 10;
       y = unpack iota 3
   in (sum for i: x.i) + (sum for j: y.j)

:t let x = unpack iota 10;
       y = unpack iota 3
   in sum for i: x.i + y.i

:t let x = unpack iota 10
   in x

x = unpack iota 10

:t x

-- yields 'compiler bug'. Should be a different sort of error
-- :t unpack x = iota 3
--    in let y = for i j: x.i * x.i
--       in sum (for i: sum (y.i))
