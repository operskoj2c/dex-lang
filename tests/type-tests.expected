:t lam x: x
A a. (a -> a)

:t lam x: sum for i: x.i
A a. (a=>Int -> Int)

:t let flip f x y = f y x in flip
A a b c. ((c -> (b -> a)) -> (b -> (c -> a)))

:t lam x: for i j: x.j.i
A a b c. (c=>b=>a -> b=>c=>a)

:t lam f x: f x
A a b. ((b -> a) -> (b -> a))

:t let id x = x
   in id (id) (id 1)
Int

:t let x = unpack iota 10
   in sum x
Int

:t let x = unpack iota 10;
       y = unpack iota 3
   in (sum for i: x.i) + (sum for j: y.j)
Int

:t let x = unpack iota 10;
       y = unpack iota 3
   in sum for i: x.i + y.i
Type error: existential variable leaked scope

:t let x = unpack iota 10
   in x
Type error: existential variable leaked scope

:t x
x=>Int

:t (lam (x, y): x + y) (1, 2)
Int

:t let f (x, y) = x + 2 * y;
       z.i = (x.i, x.i * x.i)
   in sum (for i: f z.i)
Int

